EXEMPLOS

5.2 - Cáculo da quantidade de bits necessários para uma determinada memória cache. Considere um sistema de computção com uma memória cache de 32 KB de capacidade, constituída de linhas com 8 bytes de largura. a MP possuiuma capacidade de 16 MB. 

> R = Para se saber o o total de bits necessários é preciso somar os bits de dados e os bits usados na coluna tag.

> bits de dados = quantidade de linhas x largura em bits de cada linha.

> total de bits de tag = quantidade de linhas x largura do campo tag.


In [16]:
def TOTAL_DE_BITS_DADOS(Q_linhas, Largura_bits):
    x = Q_linhas * Largura_bits
    return x

a = int(input(print(" Digite a quantidade de linhas em bits: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("total de bits de dados =",TOTAL_DE_BITS_DADOS(a, b), "bits")

 Digite a quantidade de linhas em bits: 


None 32768


 Digite a largura em bits da linha: 


None 8


total de bits de dados = 262144 bits


In [4]:
def QUANT_LINHAS(capacidade, Largura_linhas):
    x = capacidade / Largura_linhas
    return x

a = int(input(print(" Digite a capacidade cache: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("quantidade de linhas =",QUANT_LINHAS(a, b), "bytes")

 Digite a capacidade cache: 


None 32


 Digite a largura em bits da linha: 


None 8


quantidade de linhas = 4.0 bytes


> Largura do campo tag = quant de blocos / quant de linhas

In [1]:
def QUANT_BLOCOS(capacidadeMP, Largura_linha):
    x = capacidadeMP / Largura_linha
    return x

a = int(input(print(" Digite a capacidade MP: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("quantidade de linhas =",QUANT_BLOCOS(a, b), "Mbytes")

 Digite a capacidade MP: 


None 16


 Digite a largura em bits da linha: 


None 8


quantidade de linhas = 2.0 Mbytes


> QUANT_BLOCOS = 2 MB = 2**21 = 2097152

> QUANT_LINHAS = 4KB = 2**12 = 4096

In [8]:
def QUANT_BLOCOS_POR_LINHA(QUANT_BLOCOS, QUANT_LINHAS):
    x = QUANT_BLOCOS / QUANT_LINHAS
    return x

a = int(input(print(" Digite a Quantidade de blocos ")))
b = int(input(print(" Digite a Quantidade de linhas ")))
print("quantidade de blocos por linhas =", QUANT_BLOCOS_POR_LINHA(a,b),)

 Digite a Quantidade de blocos 


None 2097152


 Digite a Quantidade de linhas 


None 4096


quantidade de blocos por linhas = 512.0


In [14]:
import math
def campo_TAG(N):
    E = math.log( N, 2)
    return E
a = int(input(print(" Digite a Quantidade de blocos por linha: ")))
print("Campo TAG =",campo_TAG(a))

 Digite a Quantidade de blocos por linha: 


None 512


Campo TAG = 9.0


> QUANTIDADE DE LINHAS = 4KB campo TAG = 9 bits

In [16]:
def TOTAL_BITS_TAG(quant_linhas,campotag):
    x = quant_linhas * campotag
    return x

a = int(input(print(" Digite a Quantidadede linhas ")))
b = int(input(print(" Digite o campo tag ")))
print("total de bits dos tags =", TOTAL_BITS_TAG(a,b),)

 Digite a Quantidadede linhas 


None 4096


 Digite o campo tag 


None 9


total de bits dos tags = 36864


> total de bits cache = total bits de dados + total bits dos tags tag

In [18]:
def TOTAL_BITS_CACHE(TD,TT):
    x = TD + TT
    return x

a = int(input(print(" Digite o total bits de dados ")))
b = int(input(print(" Digite o total de bits da tag ")))
print("total de bits da cache =", TOTAL_BITS_CACHE(a,b))

 Digite o total bits de dados 


None 262144


 Digite o total de bits da tag 


None 36864


total de bits da cache = 299008


>299008 = 292KB

> FIM

5.3 - Cálculo de formato de endereço para memórias cache com mapeamento direto. Considere uma MP com 64MB de capacidade associada a uma memória cache que possui 2K linhas, cada uma com largura de 16 bytes. Determine o formato do endereço para ser interpretado pelo sistema de controle da cache.

>R = capacMP = 64*(2**23)

>linhas = 2K = 2**11

>largura da linha= 16*(2**3)

In [13]:
a = int(input(print(" Digite a capacidade MP: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("total de blocos =",QUANT_BLOCOS(a, b), "Mbytes")

 Digite a capacidade MP: 


None 64


 Digite a largura em bits da linha: 


None 16


total de blocos = 4.0 Mbytes


> total de blocos = 4 MB = 4194304 bits

In [15]:
byte = int(math.log2((16 * (2**3))/(2**3)))
print("Byte:",byte,"bits")

linha = int(math.log2(2048))
print("Linha:",linha,"bits")

Tag = int(math.log2(4194304/2048))
print("Tag:",Tag,"bits")

total = byte+linha+Tag
print("total de bits do endereço:",total,"bits")

Byte: 4 bits
Linha: 11 bits
Tag: 11 bits
total de bits do endereço: 26 bits


> FIM


5.4 - Seja uma MP constituída de blocos com largura de 32 bytes, associada a uma cache com 128KB. Em dado instante o processador realiza um acesso, colocando o seguinte endereço (expresso em algarismos hexadecimais): 3FC92B6. Determine qual deverá ser o valor binário da linha que será localizada pelo sistema de controle da cache.

In [15]:
import math
largurablocos = 32*(2**3)

capacidadecache= 128*(2**13)

Endereço = 7*4 

bytecamp = int(math.log2(largurablocos/2**3))
print ("O campo BYTE = ",bytecamp," bits")
print(" ")
a = int(input(print(" Digite a capacidade cache: ")))
b = int(input(print(" Digite a largura dos blocos: ")))
print("quantidade de linhas =",QUANT_BLOCOS(a, b),)
print(" ")
linhacamp = int(math.log2(QUANT_BLOCOS(a, b)))
print("O campo linha do endereço tem ",linhacamp, "bits")
print(" ")
tagcamp = Endereço - linhacamp - bytecamp
print(" O campo tag terá ",tagcamp,'bits')
print(" ")
print("O endereço 3FC92B6 em bits será 0011 1111 1100 1001 0010 1011 0110")
print("Dividindo pelos campos de endereço, vamos ter:")
print("Bytes 10110 bits")
print("Blocos: 010010010101 bits")
print("tag: 00111111110 bits")
print(" ")
print("O endereço da linha desejada é: 010010010101")

O campo BYTE =  5  bits
 
 Digite a capacidade cache: 


None 1048576


 Digite a largura dos blocos: 


None 256


quantidade de linhas = 4096.0
 
O campo linha do endereço tem  12 bits
 
 O campo tag terá  11 bits
 
O endereço 3FC92B6 em bits será 0011 1111 1100 1001 0010 1011 0110
Dividindo pelos campos de endereço, vamos ter:
Bytes 10110 bits
Blocos: 010010010101 bits
tag: 00111111110 bits
 
O endereço da linha desejada é: 010010010101


> FIM

5.5 - Cálculo da quantidade de bits necessários para uma determinada memória cache. Considere um sistema de computação com uma memória cache de 32KB de capacidade, constituída de linhas com 8 bytes de largura. A MP possui uma capacidade de 16MB.

> R = Para se saber o o total de bits necessários é preciso somar os bits de dados e os bits usados no campo bloco.

> bits de dados = quantidade de linhas x largura em bits de cada linha.

> total de bits do campo bloco = quantidade de linhas x largura do bloco.



> capacache = 32*(2**13) = 262144

> larguralinhas = 8*8 = 64

> capacMP = 16*(2**23) = 134217728

> bits de dados = capacache 

In [5]:
a = int(input(print(" Digite a capacidade cache: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("quantidade de linhas =",QUANT_LINHAS(a, b), "Kbytes")

 Digite a capacidade cache: 


None 32768


 Digite a largura em bits da linha: 


None 8


quantidade de linhas = 4096.0 Kbytes


In [6]:
a = int(input(print(" Digite a capacidade MP: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("quantidade de blocos =",QUANT_BLOCOS(a, b), "Mbytes")

 Digite a capacidade MP: 


None 134217728


 Digite a largura em bits da linha: 


None 64


quantidade de blocos = 2097152.0 Mbytes


In [24]:
qtdblocos = 2097152
largurablocos = math.log2(qtdblocos)
print(" largura do campo blocos = ",largurablocos,"bits")

 largura do campo blocos =  21.0 bits


In [10]:
a = int(input(print(" Digite a Quantidadede linhas ")))
b = int(input(print(" Digite a largura dos blocos ")))
print("total de bits do cache =", TOTAL_BITS_TAG(a,b),)

 Digite a Quantidadede linhas 


None 4096


 Digite a largura dos blocos 


None 21


total de bits do cache = 86016


In [13]:
a = int(input(print(" Digite o total bits de dados ")))
b = int(input(print(" Digite o total de bits do campo bloco ")))
print("total de bits da cache =", TOTAL_BITS_CACHE(a,b))

 Digite o total bits de dados 


None 262144


 Digite o total de bits do campo bloco 


None 86016


total de bits da cache = 348160


>R = 348160 = 340 Kbits

> FIM

5.6 - Cálculo de formato de endereço para memórias cache com mapeamento associativo completo. Considere uma MP com 64MB de capacidade associada a uma memória cache que possui 2K linhas, cada uma com largura de 16 bytes. Determine o formato do endereço para ser interpretado pelo sistema de controle da cache.

>capacMP = 64*(2**23) = 536870912

>linhas = 2**21 = 

>largura bloco = 16*8 = 128

In [16]:
a = int(input(print(" Digite a capacidade MP: ")))
b = int(input(print(" Digite a largura dos blocos: ")))
print("O total de blocos =",QUANT_BLOCOS(a, b))

 Digite a capacidade MP: 


None 536870912


 Digite a largura dos blocos: 


None 128


O total de blocos = 4194304.0


In [18]:
import math
x = 4194304
largurabloco = 128
byte = int(math.log2(largurabloco/8))
print("bytes:",byte,"bits")
bloco = int(math.log2(x))
print("bloco:",bloco,"bits")

bytes: 4 bits
bloco: 22 bits


> FIM

5.7 - Seja uma MP constituída de blocos com largura de 32 bytes, associada a uma cache com 64KB. Em dado instante o processador realiza um acesso, colocando o seguinte endereço (expresso em algarismos hexadecimais): 3FC92B6. Determine qual deverá ser o valor binário do campo bloco que será localizado pelo sistema de controle da cache.

>largura blocos = 32*8

>capacidade cache = 64*(2**13)

In [4]:
import math
largblocos = 32*8
capacache = 64*2**13

E = len("3FC92B6")*4
bytecamp = math.log2(largblocos/8)
qtdbloco = 2**E/2**bytecamp
campbloco = math.log2(qtdbloco)

inteiro = int("3FC92B6", base=16)
bi = bin(inteiro)[::-1]

#bytes
byte = ""
for i in range(int(bytecamp)):
    byte += bi[i]
byte = byte[::-1]
print("Bytes",byte,"bits")

#bloco
bloco = ""
for j in range(int(campbloco)):
    bloco += bi[j+int(bytecamp)]
bloco = bloco[::-1]
print("Blocos:",bloco,"bits")


Bytes 10110 bits
Blocos: 0b111111110010010010101 bits


> FIM

5.8 - Cálculo da quantidade de bits necessários para uma determinada memória cache, que funciona com mapeamento por conjunto de quatro. Considere um sistema de computação com uma memória cache de 32KB de capacidade, constituída de linhas com 8 bytes de largura e conjunto de 4. A MP possui uma capacidade de 16MB.

> capacidade cache = 32*(2**13) = 262144

> largura linha = 8*8

>capacidade MP = 16*(2**23) = 134217728

>quantidade de linhas conjunto = 4

R = Para se saber o o total de bits necessários é preciso somar os bits de dados e os bits usados no campo bloco.

bits de dados = quantidade de linhas x largura em bits de cada linha.

total de bits de tag = quantidade de linhas x largura do campo tag

>bits de dados = capacidade cache 

In [7]:
a = int(input(print(" Digite a capacidade cache: ")))
b = int(input(print(" Digite a largura em bits da linha: ")))
print("quantidade de linhas =",QUANT_LINHAS(a, b), "bytes")

 Digite a capacidade cache: 


None 262144


 Digite a largura em bits da linha: 


None 64


quantidade de linhas = 4096.0 bytes


In [21]:
a = int(input(print(" Digite a quantidade de linhas: ")))
b = int(input(print(" Digite a quantidade de linhas do conjunto: ")))
print("quantidade de conjuntos =",QUANT_BLOCOS(a, b),)

 Digite a quantidade de linhas: 


None 4096


 Digite a quantidade de linhas do conjunto: 


None 4


quantidade de conjuntos = 1024.0


In [11]:
a = int(input(print(" Digite a capacidade MP: ")))
b = int(input(print(" Digite a largura da linha: ")))
print("O total de blocos =",QUANT_BLOCOS(a, b))

 Digite a capacidade MP: 


None 134217728


 Digite a largura da linha: 


None 64


O total de blocos = 2097152.0


In [22]:
qtdblocos = 2097152.0
qtdconjuntos = 1024
camptag = qtdblocos/qtdconjuntos
print(" Campo tag = ",camptag)
print(" ")
a = int(input(print(" Digite o campo da tag: ")))
print("Campo LARG =",campo_TAG(a))

 Campo tag =  2048.0
 
 Digite o campo da tag: 


None 2048


Campo LARG = 11.0


In [24]:
a = int(input(print(" Digite a Quantidadede linhas ")))
b = int(input(print(" Digite a largura do campo tag ")))
print("total de bits dos tags =", TOTAL_BITS_TAG(a,b),)

 Digite a Quantidadede linhas 


None 4096


 Digite a largura do campo tag 


None 11


total de bits dos tags = 45056


In [25]:
a = int(input(print(" Digite o total bits de dados ")))
b = int(input(print(" Digite o total de bits da tag ")))
print("total de bits da cache =", TOTAL_BITS_CACHE(a,b))

 Digite o total bits de dados 


None 262144


 Digite o total de bits da tag 


None 45056


total de bits da cache = 307200


R = 307200 / 2**10 = 300 bits

>FIM

5.9 - Cálculo de formato de endereço para memórias cache com mapeamento associativo por conjunto. Considere uma MP com 64MB de capacidade associada a uma memória cache que funciona com mapeamento associativo por conjunto de 4 e que possui 32KB, com linhas de largura de 16 bytes. Determine o formato do endereço para ser interpretado pelo sistema de controle da cache.

>capacidade MP = 64*(2**23) = 536870912

>conjunto = 4

>largura bloco = 16*8 = 128

>capacidade cache = 32*2**13 = 262144

In [3]:
a = int(input(print(" Digite a capacidade MP: ")))
b = int(input(print(" Digite a largura do bloco: ")))
print("Quantidade de blocos da MP =",QUANT_BLOCOS(a, b))

 Digite a capacidade MP: 


None 536870912


 Digite a largura do bloco: 


None 128


Quantidade de blocos da MP = 4194304.0


In [5]:
a = int(input(print(" Digite a capacidade cache: ")))
b = int(input(print(" Digite a largura do bloco: ")))
print("quantidade de linhas de cache =",QUANT_LINHAS(a, b), "bytes")

 Digite a capacidade cache: 


None 262144


 Digite a largura do bloco: 


None 128


quantidade de linhas de cache = 2048.0 bytes


In [6]:
a = int(input(print(" Digite a quantidade de linhas de cache: ")))
b = int(input(print(" Digite a quantidade de linhas do conjunto: ")))
print("quantidade de conjuntos =",QUANT_BLOCOS(a, b),)

 Digite a quantidade de linhas de cache: 


None 2048


 Digite a quantidade de linhas do conjunto: 


None 4


quantidade de conjuntos = 512.0


In [7]:
a = int(input(print(" Digite a Quantidade de blocos da MP: ")))
b = int(input(print(" Digite a quantidade de conjuntos: ")))
print("quantidade de bloco de conjuntos =",QUANT_BLOCOS(a, b),)

 Digite a Quantidade de blocos da MP: 


None 4194304


 Digite a quantidade de conjuntos: 


None 512


quantidade de bloco de conjuntos = 8192.0


In [8]:
import math
largurabloco = 128
qtdconjunto = 512
qtdblococonjunto = 8192

bytecamp = math.log2(largurabloco/8)

conjuntocamp = math.log2(qtdconjunto)

tagcamp = math.log2(qtdblococonjunto)

print("Byte:",int(bytecamp),"bits")
print("Conjunto:",int(conjuntocamp),"bits")
print("Tag:",int(tagcamp),"bits")


Byte: 4 bits
Conjunto: 9 bits
Tag: 13 bits


>FIM

5.10 - Seja uma MP constituída de blocos com largura de 32 bytes, associada a uma cache com 64KB; a cache usa mapeamento por conjunto de 4. Em dado instante o processador realiza um acesso, colocando o seguinte endereço (expresso em algarismos hexadecimais): 3FC92B6. Determine qual deverá ser o valor binário do conjunto que será localizado pelo sistema de controle da cache.

>largura do bloco = 32*8 = 256

>capacidade cache = 64*(2**13) = 524288

>conjunto = 4

In [15]:
largbloco = 32*8
capacache = 64*2**13
conjunto = 4

E = len("3FC92B6")*4
bytecamp = math.log2(largbloco/8)
print("Bytes:",int(bytecamp),"bits")

qtdlinhas = capacache/(2**bytecamp*8)

qtdconjunto=qtdlinhas/conjunto

conjuntocamp = math.log2(qtdconjunto)
print("Conjunto:",int(conjuntocamp),"bits")

tagcamp = E - conjuntocamp - bytecamp
print("TAG:",int(tagcamp),"bits")

print("Conjunto: 010010101")
print("tag: 0b111111110010")
print("O endereço da linha desejada é: 010010101")

Bytes: 5 bits
Conjunto: 9 bits
TAG: 14 bits
Conjunto: 010010101
tag: 0b111111110010
O endereço da linha desejada é: 010010101
